synapse/storage/databases/main/receipts.py

```diff
@@ -941,34 +941,39 @@ async def _background_receipts_linearized_unique_index(
         receipts."""
 
         def _remote_duplicate_receipts_txn(txn: LoggingTransaction) -> None:
+            if isinstance(self.database_engine, PostgresEngine):
+                ROW_ID_NAME = "ctid"
+            else:
+                ROW_ID_NAME = "rowid"
+
             # Identify any duplicate receipts arising from
             # https://github.com/matrix-org/synapse/issues/14406.
             # We expect the following query to use the per-thread receipt index and take
             # less than a minute.
```

Comment:
I'm not sure what I was thinking when I wrote this comment, but postgres didn't use the index at all when I tested just now.

<details>
<summary>Query plan</summary>

```
matrix=> EXPLAIN ANALYZE SELECT MAX(stream_id), MAX(ctid), room_id, receipt_type, user_id
                FROM receipts_linearized
                WHERE thread_id IS NULL
                GROUP BY room_id, receipt_type, user_id
                HAVING COUNT(*) > 1;
                                                                             QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize GroupAggregate  (cost=1227914.68..1611643.58 rows=463693 width=74) (actual time=37505.556..37790.249 rows=0 loops=1)
   Group Key: room_id, receipt_type, user_id
   Filter: (count(*) > 1)
   Rows Removed by Filter: 13165939
   ->  Gather Merge  (cost=1227914.68..1552522.68 rows=2782160 width=82) (actual time=18205.220..29774.135 rows=13165939 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Sort  (cost=1226914.66..1230392.36 rows=1391080 width=82) (actual time=16726.364..22540.347 rows=4388646 loops=3)
               Sort Key: room_id, receipt_type, user_id
               Sort Method: external merge  Disk: 401216kB
               Worker 0:  Sort Method: external merge  Disk: 455888kB
               Worker 1:  Sort Method: external merge  Disk: 429192kB
               ->  Partial HashAggregate  (cost=884550.89..1037422.93 rows=1391080 width=82) (actual time=5403.068..11355.633 rows=4388646 loops=3)
                     Group Key: room_id, receipt_type, user_id
                     Planned Partitions: 32  Batches: 193  Memory Usage: 16657kB  Disk Usage: 422968kB
                     Worker 0:  Batches: 197  Memory Usage: 16657kB  Disk Usage: 456432kB
                     Worker 1:  Batches: 201  Memory Usage: 16657kB  Disk Usage: 437264kB
                     ->  Parallel Seq Scan on receipts_linearized  (cost=0.00..415075.68 rows=5472935 width=74) (actual time=261.776..1775.731 rows=4388646 loops=3)
                           Filter: (thread_id IS NULL)
                           Rows Removed by Filter: 260882
 Planning Time: 0.226 ms
 JIT:
   Functions: 40
   Options: Inlining true, Optimization true, Expressions true, Deforming true
   Timing: Generation 10.377 ms, Inlining 284.499 ms, Optimization 552.845 ms, Emission 392.178 ms, Total 1239.899 ms
 Execution Time: 38162.669 ms
(26 rows)
```
</details>

```suggestion
            # The following query takes less than a minute on matrix.org.
```

Reply:
Well, regardless it has been faster on any of the servers I've run it on! :)